oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,399 @@
+    } catch (NoSuchAlgorithmException e) {
+      // Will not occur as HmacSHA256 is supported. We may allow other algorithms in the future.
+      throw new RuntimeException("HmacSHA256 must be supported by the JVM.", e);
+    } catch (InvalidKeyException e) {
+      throw new SigningException("Invalid key used when calculating the AWS V4 Signature", e);
+    }
+  }
+
+  private static String getHexEncodedSha256Hash(byte[] bytes) {
+    try {
+      MessageDigest digest = MessageDigest.getInstance("SHA-256");
+      return BaseEncoding.base16().lowerCase().encode(digest.digest(bytes));
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException("Failed to compute SHA-256 hash.", e);
+    }
+  }
+
+  static Builder newBuilder(
+      AwsSecurityCredentials awsSecurityCredentials, String httpMethod, String url, String region) {
+    return new Builder(awsSecurityCredentials, httpMethod, url, region);
+  }
+
+  static class Builder {
+
+    private final AwsSecurityCredentials awsSecurityCredentials;
+    private final String httpMethod;
+    private final String url;
+    private final String region;
+
+    @Nullable private String requestPayload;
+    @Nullable private Map<String, String> additionalHeaders;
+    @Nullable private AwsDates dates;
+
+    private Builder(
+        AwsSecurityCredentials awsSecurityCredentials,
+        String httpMethod,
+        String url,
+        String region) {
+      this.awsSecurityCredentials = awsSecurityCredentials;
+      this.httpMethod = httpMethod;
+      this.url = url;
+      this.region = region;
+    }
+
+    Builder setRequestPayload(String requestPayload) {
+      this.requestPayload = requestPayload;
+      return this;
+    }
+
+    Builder setAdditionalHeaders(Map<String, String> additionalHeaders) {
+      if (additionalHeaders.containsKey("date") && additionalHeaders.containsKey("x-amz-date")) {
+        throw new IllegalArgumentException("One of {date, x-amz-date} can be specified, not both.");
+      }
+      try {
+        if (additionalHeaders.containsKey("date")) {
+          this.dates = AwsDates.fromDateHeader(additionalHeaders.get("date"));
+        }
+        if (additionalHeaders.containsKey("x-amz-date")) {
+          this.dates = AwsDates.fromXAmzDate(additionalHeaders.get("x-amz-date"));
+        }
+      } catch (ParseException e) {
+        throw new IllegalArgumentException("The provided date header value is invalid.", e);
+      }
+
+      this.additionalHeaders = additionalHeaders;
+      return this;
+    }
+
+    AwsRequestSigner build() {
+      return new AwsRequestSigner(
+          awsSecurityCredentials,
+          httpMethod,
+          url,
+          region,
+          requestPayload,
+          additionalHeaders,
+          dates);
+    }
+  }
+
+  static final class AwsDates {
+
+    private static final String X_AMZ_DATE_FORMAT = "yyyyMMdd'T'HHmmss'Z'";
+    private static final String CUSTOM_DATE_FORMAT = "E, dd MMM yyyy HH:mm:ss z";
```

Comment:
why *CUSTOM*_DATE_FORMAT? Is this the HTTP-date format defined by RFC 7231? If so, rename to HTTP_DATE_FORMAT for clarity.